## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [88]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 18 columns]

In [89]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [90]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [91]:
# Use this space to do any additional research
# and familiarize yourself with the data.
category_df = df[["first", "last", "client_id", "category", "subcategory", "qty"]].copy()
category_df

first       last  client_id     category         subcategory   qty
0         Donald    Harding      58515        decor            wall art   105
1        Tiffany      Myers      37609  consumables                pens    21
2        Shannon     Watson      57113     software  project management    39
3         Nathan      Baker      46554  consumables                pens    29
4      Christina   Schwartz      92089  consumables                misc    20
...          ...        ...        ...          ...                 ...   ...
54634     Lauren      Reese      29043     software           logistics    33
54635    Derrick      Moore      41908  consumables    kitchen supplies    47
54636     Monica  Gutierrez      35176  consumables   bathroom supplies   475
54637      Wanda    Solomon      24485    furniture              stands   112
54638     Dwayne       Bell      24007     software  project management  1031

[54639 rows x 6 columns]

In [92]:
# What three item categories had the most entries?
Three_categories = pd.DataFrame(category_df)
Categories_count = Three_categories["category"].value_counts()
Categories_count.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [93]:
# For the category with the most entries,
# which subcategory had the most entries?
categories_dropped = category_df.drop(category_df[category_df["category"] != "consumables"].index)
subcat_count = categories_dropped["subcategory"].value_counts()
subcat_count.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [94]:
# Which five clients had the most entries in the data?
Client_count = category_df["client_id"].value_counts()
top_5 = Client_count.head(5)
top_5

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [95]:
# Store the client ids of those top 5 clients in a list.
top_5_client = pd.Series([33615, 66037, 46820, 38378, 24741])
top_5_list = top_5_client.to_list()
top_5_list

[33615, 66037, 46820, 38378, 24741]

In [96]:
# How many total units (the qty column) did the
# client with the most entries order order?
clients_droppes = category_df.drop(category_df[category_df["client_id"] != 33615].index)
client_tot_units = clients_droppes["qty"].sum()
client_tot_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [97]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
transform_df = pd.DataFrame(df)
subtotal_calc = transform_df[["unit_price", "qty"]].copy()
subtotal_calc["line_subtotal"] = subtotal_calc["unit_price"] * subtotal_calc["qty"]
subtotal_calc

unit_price   qty  line_subtotal
0         1096.80   105      115164.00
1           24.95    21         523.95
2           13.52    39         527.28
3           36.42    29        1056.18
4          195.10    20        3902.00
...           ...   ...            ...
54634       83.13    33        2743.29
54635      206.59    47        9709.73
54636       65.66   475       31188.50
54637        1.48   112         165.76
54638        3.01  1031        3103.31

[54639 rows x 3 columns]

In [98]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
data = pd.DataFrame(df)
shipping_info = data[["first", "last","client_id", "order_id", "unit_price", "unit_cost", "unit_weight", "qty",]].copy()
def shipping_weight(row):
        total_weight = row["unit_weight"] * row["qty"]
        if total_weight > 50:
            return 7 * total_weight
        elif total_weight <= 50:
            return 10 * total_weight

shipping_info["shipping_price"] = shipping_info.apply(shipping_weight, axis = 1)
shipping_info

first       last  client_id  order_id  unit_price  unit_cost  \
0         Donald    Harding      58515   8953482     1096.80     762.71   
1        Tiffany      Myers      37609   8069089       24.95      15.09   
2        Shannon     Watson      57113   1902144       13.52       7.86   
3         Nathan      Baker      46554   9031802       36.42      24.85   
4      Christina   Schwartz      92089   1322274      195.10     108.17   
...          ...        ...        ...       ...         ...        ...   
54634     Lauren      Reese      29043   9021716       83.13      51.60   
54635    Derrick      Moore      41908   6290153      206.59     175.46   
54636     Monica  Gutierrez      35176   8692622       65.66      57.31   
54637      Wanda    Solomon      24485   7592730        1.48       1.22   
54638     Dwayne       Bell      24007   7489403        3.01       1.61   

       unit_weight   qty  shipping_price  
0             7.50   105         5512.50  
1             1.49    21          312.90  
2             1.68    39          458.64  
3             1.23    29          356.70  
4            46.43    20         6500.20  
...            ...   ...             ...  
54634         2.25    33          519.75  
54635        11.70    47         3849.30  
54636         4.16   475        13832.00  
54637        18.04   112        14143.36  
54638         2.07  1031        14939.19  

[54639 rows x 9 columns]

In [99]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
data = shipping_info.copy()
#def total_price_per_unit(row):
    #sub_total = ((row["unit_price"]) + (row["shipping_price"]/row["qty"])) 
    #sales_tax = sub_total * (9.25/100)
    #total_price_unit = sub_total + sales_tax
    #return total_price_unit
#shipping_info["Total_price_per_unit"] = shipping_info.apply(total_price_per_unit, axis = 1)
def total_price(row):
    sub_total_tot = ((row["unit_price"] * row["qty"]) + row["shipping_price"]) 
    sales_tax = sub_total_tot * (9.25/100)
    total_p = sub_total_tot + sales_tax
    return total_p
shipping_info["total_price"] = shipping_info.apply(total_price, axis = 1)
shipping_info

first       last  client_id  order_id  unit_price  unit_cost  \
0         Donald    Harding      58515   8953482     1096.80     762.71   
1        Tiffany      Myers      37609   8069089       24.95      15.09   
2        Shannon     Watson      57113   1902144       13.52       7.86   
3         Nathan      Baker      46554   9031802       36.42      24.85   
4      Christina   Schwartz      92089   1322274      195.10     108.17   
...          ...        ...        ...       ...         ...        ...   
54634     Lauren      Reese      29043   9021716       83.13      51.60   
54635    Derrick      Moore      41908   6290153      206.59     175.46   
54636     Monica  Gutierrez      35176   8692622       65.66      57.31   
54637      Wanda    Solomon      24485   7592730        1.48       1.22   
54638     Dwayne       Bell      24007   7489403        3.01       1.61   

       unit_weight   qty  shipping_price    total_price  
0             7.50   105         5512.50  131839.076250  
1             1.49    21          312.90     914.258625  
2             1.68    39          458.64    1077.117600  
3             1.23    29          356.70    1543.571400  
4            46.43    20         6500.20   11364.403500  
...            ...   ...             ...            ...  
54634         2.25    33          519.75    3564.871200  
54635        11.70    47         3849.30   14813.240275  
54636         4.16   475        13832.00   49184.896250  
54637        18.04   112        14143.36   15632.713600  
54638         2.07  1031        14939.19   19711.431250  

[54639 rows x 10 columns]

In [100]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
data = shipping_info.copy()
def total_cost(row):
    cost =  (row["unit_cost"] * row["qty"]) + row["shipping_price"]
    return cost
shipping_info["cost"] = shipping_info.apply(total_cost, axis = 1)
shipping_cost = shipping_info
shipping_cost 

first       last  client_id  order_id  unit_price  unit_cost  \
0         Donald    Harding      58515   8953482     1096.80     762.71   
1        Tiffany      Myers      37609   8069089       24.95      15.09   
2        Shannon     Watson      57113   1902144       13.52       7.86   
3         Nathan      Baker      46554   9031802       36.42      24.85   
4      Christina   Schwartz      92089   1322274      195.10     108.17   
...          ...        ...        ...       ...         ...        ...   
54634     Lauren      Reese      29043   9021716       83.13      51.60   
54635    Derrick      Moore      41908   6290153      206.59     175.46   
54636     Monica  Gutierrez      35176   8692622       65.66      57.31   
54637      Wanda    Solomon      24485   7592730        1.48       1.22   
54638     Dwayne       Bell      24007   7489403        3.01       1.61   

       unit_weight   qty  shipping_price    total_price      cost  
0             7.50   105         5512.50  131839.076250  85597.05  
1             1.49    21          312.90     914.258625    629.79  
2             1.68    39          458.64    1077.117600    765.18  
3             1.23    29          356.70    1543.571400   1077.35  
4            46.43    20         6500.20   11364.403500   8663.60  
...            ...   ...             ...            ...       ...  
54634         2.25    33          519.75    3564.871200   2222.55  
54635        11.70    47         3849.30   14813.240275  12095.92  
54636         4.16   475        13832.00   49184.896250  41054.25  
54637        18.04   112        14143.36   15632.713600  14280.00  
54638         2.07  1031        14939.19   19711.431250  16599.10  

[54639 rows x 11 columns]

In [101]:
# Create a column for the profit of
# each line using line cost and line price
shipping_cost["Profit"] = shipping_cost["total_price"] - shipping_cost["cost"]
shipping_cost


first       last  client_id  order_id  unit_price  unit_cost  \
0         Donald    Harding      58515   8953482     1096.80     762.71   
1        Tiffany      Myers      37609   8069089       24.95      15.09   
2        Shannon     Watson      57113   1902144       13.52       7.86   
3         Nathan      Baker      46554   9031802       36.42      24.85   
4      Christina   Schwartz      92089   1322274      195.10     108.17   
...          ...        ...        ...       ...         ...        ...   
54634     Lauren      Reese      29043   9021716       83.13      51.60   
54635    Derrick      Moore      41908   6290153      206.59     175.46   
54636     Monica  Gutierrez      35176   8692622       65.66      57.31   
54637      Wanda    Solomon      24485   7592730        1.48       1.22   
54638     Dwayne       Bell      24007   7489403        3.01       1.61   

       unit_weight   qty  shipping_price    total_price      cost  \
0             7.50   105         5512.50  131839.076250  85597.05   
1             1.49    21          312.90     914.258625    629.79   
2             1.68    39          458.64    1077.117600    765.18   
3             1.23    29          356.70    1543.571400   1077.35   
4            46.43    20         6500.20   11364.403500   8663.60   
...            ...   ...             ...            ...       ...   
54634         2.25    33          519.75    3564.871200   2222.55   
54635        11.70    47         3849.30   14813.240275  12095.92   
54636         4.16   475        13832.00   49184.896250  41054.25   
54637        18.04   112        14143.36   15632.713600  14280.00   
54638         2.07  1031        14939.19   19711.431250  16599.10   

             Profit  
0      46242.026250  
1        284.468625  
2        311.937600  
3        466.221400  
4       2700.803500  
...             ...  
54634   1342.321200  
54635   2717.320275  
54636   8130.646250  
54637   1352.713600  
54638   3112.331250  

[54639 rows x 12 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [102]:
# Check your work using the totals above
confirmation = shipping_cost[["client_id","order_id", "total_price"]].copy()
Order_274 = confirmation["order_id"] == 2742071
Order_217 = confirmation["order_id"] == 2173913
Order_612 = confirmation["order_id"] == 6128929
order_274_price = confirmation.loc[Order_274, "total_price"].sum()
order_217_price = confirmation.loc[Order_217, "total_price"].sum()
order_612_price = confirmation.loc[Order_612, "total_price"].sum()
print(order_274_price)
print(order_217_price)
print(order_612_price)

152811.897075
162388.71929999997
923441.24045


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [103]:
#33615: $8377308.52
#66037: $10259514.79
#46820: $9743794.36
#38378: $12906550.87
#24741: $82268892.02
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_5_clients = [33615, 66037, 46820, 38378, 24741]
client_33615 = {33615: "total_price"}
client_66037 = {66037: "total_price"}
client_46820 = {46820: "total_price"}
client_38378 = {38378: "total_price"}
client_24741 = {24741: "total_price"}
for client in top_5_clients:
   total_spent = shipping_cost.loc[shipping_cost["client_id"] == client, 'total_price'].sum()
   if client == 33615:
      client_33615[client] = total_spent
   elif client == 66037:
      client_66037[client] = total_spent
   elif client == 46820:
      client_46820[client] = total_spent
   elif client == 38378:
      client_38378[client] = total_spent
   elif client == 24741:
      client_24741[client] = total_spent
print(client_33615)
print(client_66037)
print(client_46820)
print(client_38378)
print(client_24741)
    #if shipping_cost[shipping_cost["client_id"].isin(top_5_clients)] == True:
   # filtered_totals = shipping_cost[shipping_cost["client_id"].isin(top_5_clients)]
   # filtered_totals.head(30)





{33615: 8377308.572500001}
{66037: 10259514.801275}
{46820: 9743794.320474999}
{38378: 12906550.880874999}
{24741: 82268891.97985001}


In [109]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_data_frame = shipping_cost[["client_id", "qty", "shipping_price", "total_price", "cost", "Profit"]].copy()

summary_pt_2 = summary_data_frame.loc[summary_data_frame["client_id"].isin(top_5_clients)]
final_form = summary_pt_2.pivot_table(index = "client_id", aggfunc="sum").reset_index().sort_values(by = "Profit", ascending = False)
final_form

client_id        Profit         cost     qty  shipping_price   total_price
0      24741  3.657999e+07  45688899.71  239862      5126448.37  8.226889e+07
2      38378  3.271830e+06   9634720.98   73667      3429455.40  1.290655e+07
4      66037  3.255032e+06   7004482.98   43018      1395151.85  1.025951e+07
3      46820  2.736603e+06   7007191.64   75768      1601448.84  9.743794e+06
1      33615  2.201995e+06   6175313.91   64313      1828984.89  8.377309e+06

In [110]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
updated_names = final_form.rename(columns = {"client_id" : "Client ID", "Profit": "Total Profit", "cost": "Total Cost", "qty":"Units", "shipping_price":"Shipping", "total_price":"Total Revenue"})
updated_order = updated_names[["Client ID", "Units", "Shipping", "Total Revenue", "Total Cost", "Total Profit"]]
mil_scale =["Shipping", "Total Revenue", "Total Cost", "Total Profit"]
def scale_to_millions(row, mil_scale):
        return f"${(row[mil_scale]/1_000_000).round(2)}M"
data_formated = updated_order.copy()
for columns in mil_scale:
    data_formated[columns]= updated_order.apply(scale_to_millions, axis = 1, mil_scale = columns)
data_formated

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
3      46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
1      33615   64313   $1.83M        $8.38M     $6.18M        $2.2M

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
sorted_order =  data_formated.sort_values("Total Profit", ascending = False)
sorted_order

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
3      46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
1      33615   64313   $1.83M        $8.38M     $6.18M        $2.2M

# SUMMARY
Overall this business has shown to be a very lucritive striclty number wise.
    The majority of the business\'s profit\'s come from the product categories consumables, furniture, and software.
    It appears that client 24741 has been one of the largest benefactors to the business by far spending $82.27 million themselves.
    This blows all other customers out of the water especially the second largest customer which by comparison spent $12.91 million.
    It would be wise of this company to continue it's good relations with said customer is they want their business to keep growing.